LLama/Control/IGenerationControl.cs

```diff
@@ -0,0 +1,35 @@
+﻿using LLama.Abstractions;
+using System;
+using System.Collections.Generic;
+using System.Text;
+
+namespace LLama.Control
+{
+    /// <summary>
+    /// Control the text generation of LLama Executors.
+    /// </summary>
+    public interface IGenerationControl
+    {
+        /// <summary>
+        /// Use the last output text to determine if the generation should stop.
+        /// This method will be called after the overload with output id.
+        /// The text will be returned even if this returns true but the generation will be stopped.
+        /// </summary>
+        /// <param name="context">The LLamaContext used in the current generation.</param>
+        /// <param name="inferenceParams">The inference params used in the current generation.</param>
+        /// <param name="lastOutputText">The last output text generated by the model.</param>
+        /// <returns></returns>
+        bool ShouldStopGeneration(LLamaContext context, IInferenceParams inferenceParams, string lastOutputText);
```

Comment:
Just a thought - is it possible to decouple this interface from the `IInferenceParams` somehow? It feels a bit odd that `GenerationControl` is overriding parts of `IInferenceParams` and is _also_ configured by it.

The obvious way to do that would be to remove the `inferenceParams` parameter here, that would introduce problems in `DefaultGenerationControl` though.

Reply:
Yeah I also thought of that and I could come up with the following alternatives:

1. Change the API of `ILLamaExecutor` to `InferAsync(string text, IInferenceParams? inferenceParams = null, IGenerationControl? control = null, CancellationToken token = default);` or similar formats. It will introduce some break changes though. 
2. Use generic type. Change the `ILLamaExecutor` to `ILLamaExecutor<TControl>`, or change the method `InferAsync`.
3. Use chained-call. For example, in the sight of users, they could do `var executor = new StatelessExecutor().WithControl()`.

Each of them has advantages and disadvantages. The most important one I think is that whether we should make it class level or method level. Actually I can't think up a case that the user must use different control strategies in different calls of an executor, though it does have such flexibility. If to get a compromise of it, I'd like to suggest the following proposal:

1. Use the option 3 above, the chained-call for `ILLamaExecutor`, taking `WithGenerationControl` as an extension method.
2. Define a static method to execute the generation, just like what I did in #362. In this static method, we could introduce any break change, like option 1 above. Since dotnet 7, we could define a static method in the interface but in older version it's not supported. Therefore if taking this approach, some parts of the design may look a little awkward.

BTW there's one thing related with it, the sampling pipeline. Currently `InferenceParams` contains many params of sampling, while specifying `SamplingPipeline` will totally override them. It's completely okay in the master branch because we don't want to introduce substantial break changes now. However I'm wondering if we should refactor it before v1.0.0.
For example, we could add many kinds of sampling pipelines as we discussed [here](https://github.com/SciSharp/LLamaSharp/pull/348#issuecomment-1848427031), along with a class named `CustomSamplingPipeline`, which accepts all the parameters related with samping in `InferenceParams` now.

Reply:
I was thinking more along the lines of removing the `IInferenceParams inferenceParams` from the `ShouldStopGeneration` methods.

Reply:
> It's completely okay in the master branch because we don't want to introduce substantial break changes now. However I'm wondering if we should refactor it before v1.0.0.

Yeah this is my thinking. It feels like we're moving in a direction where we can completely get rid of the inference params in the future (which is part of why I want to avoid depending on it in this new interface, if possible).